xtrlock.c

```diff
@@ -18,196 +18,373 @@
 
 #include <X11/X.h>
 #include <X11/Xlib.h>
+#include <X11/Xos.h>
 #include <X11/Xutil.h>
 #include <X11/keysym.h>
-#include <X11/Xos.h>
 
-#include <stdlib.h>
-#include <stdio.h>
-#include <sys/types.h>
-#include <sys/stat.h>
+#include <crypt.h>
+#include <ctype.h>
 #include <errno.h>
-#include <pwd.h>
 #include <grp.h>
 #include <limits.h>
+#include <math.h>
+#include <pwd.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
-#include <crypt.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <time.h>
 #include <unistd.h>
-#include <math.h>
-#include <ctype.h>
 #include <values.h>
 
 #ifdef SHADOW_PWD
 #include <shadow.h>
 #endif
 
-Display *display;
-Window window, root;
+//#define DEBUG
+#ifdef DEBUG
+#define debug_print(...)             \
+        do {                         \
+                printf(__VA_ARGS__); \
+        } while (0)
+#else
+#define debug_print(...) \
+        do {             \
+        } while (0)
+#endif
+
+#define msleep(x) usleep(x * 1000)
+
+Display* display;
+Window window, blank_window, trans_window;
 
 #define TIMEOUTPERATTEMPT 30000
-#define MAXGOODWILL  (TIMEOUTPERATTEMPT*5)
+#define MAXGOODWILL (TIMEOUTPERATTEMPT * 5)
 #define INITIALGOODWILL MAXGOODWILL
 #define GOODWILLPORTION 0.3
+bool blank_screen = false;
+int blink_delay = 100;
 
-struct passwd *pw;
-int passwordok(const char *s) {
-#if 0
-  char key[3];
-  char *encr;
-  
-  key[0] = *(pw->pw_passwd);
-  key[1] =  (pw->pw_passwd)[1];
-  key[2] =  0;
-  encr = crypt(s, key);
-  return !strcmp(encr, pw->pw_passwd);
-#else
-  /* simpler, and should work with crypt() algorithms using longer
-     salt strings (like the md5-based one on freebsd).  --marekm */
-  return !strcmp(crypt(s, pw->pw_passwd), pw->pw_passwd);
-#endif
+struct { /*Setting correspond to the custom passwd setting. --d0048*/
+        bool enable;
+        char* pwd;
+} cust_pw_setting;
+
+struct passwd* pw;
+int passwordok(const char* s)
+{
+        /* simpler, and should work with crypt() algorithms using longer
+           salt strings (like the md5-based one on freebsd).  --marekm */
+        debug_print("%s, %i\n", s, (int)strlen(s));
+
+        if (cust_pw_setting.enable) {
+                debug_print("Entered_de: %s\n", s);
+                debug_print("Original_de: %s\n", cust_pw_setting.pwd);
+                char* enter = strdup(crypt(s, cust_pw_setting.pwd));
+                char* original = cust_pw_setting.pwd;
+                if (NULL == enter) {
+                        fprintf(stderr, "\"strdup\" or \"crypt\":%s\n", strerror(errno));
+                        return false;
+                }
+                debug_print("Entered:  %s\n", enter);
+                debug_print("Original: %s\n", original);
+                unsigned int i = strcmp(enter, original);
+                free(enter);
+                return !i;
+        } else { /*salt in the second argument seems to be automatically used by crypt*/
```

Comment:
I've removed the redundant code. If there are no more problems, I will be drawing a release and be started on the next version. (I read the article [here](https://www.jwz.org/xscreensaver/toolkits.html) and got really impressed.  Maybe I will try adding a daemon process to this program in the next version.)